lib/ddtrace/sampler.rb

```diff
@@ -247,7 +248,15 @@ def priority_assigned_upstream?(span)
 
     def priority_sample!(span)
       preserving_sampling(span) do
-        @priority_sampler.sample!(span)
+        @priority_sampler.sample!(span).tap do
+          # We want to make sure the span is tagged with the agent-derived
+          # service rate. Retrieve this from the rate by service sampler.
+          # Only do this if it was set by a RateByServiceSampler.
+          if @priority_sampler.is_a?(RateByServiceSampler)
```

Comment:
From https://github.com/DataDog/dd-trace-rb/pull/891#pullrequestreview-340722745, @brettlangdon commented:
> as long as we only set the metric when RateByServiceSampler is used, and we always set the metric regardless of whether the span is sampled by RateByServiceSampler or not

Since https://github.com/DataDog/dd-trace-rb/pull/917 was merged, the default `@priority_sampler` in this line will be a RuleSampler: https://github.com/DataDog/dd-trace-rb/blob/930d8210ab2bc6f54b35503aa41be9bb674c1c93/lib/ddtrace/tracer.rb#L409

The RuleSampler has a fallback `RateByServiceSampler` inside by default, but it only uses it if no rules match, which is also the default.

I'm thinking that one possible solution is to move this metric setting logic to `RateByServiceSampler#sample!` instead.

Reply:
My issue with moving it there is that not everything that samples by service rate is priority sampling... its a flawed assumption that could lead to incorrect metrics/sample rates.

Reply:
> not everything that samples by service rate is priority sampling

it should be, that is the only use for `RateByServiceSampler`, that is priority sampling controlled by the agent.

Reply:
it also doesn't necessarily matter, since the tag is meant to indicate that agent controlled sampling had a hand in deciding the sampling of a span.

Reply:
Would it not be possible, even if not probable, for a user to define `RateByServiceSampler` as their default sampler and set the sample rates directly for each of their services that way? Seems like a legitimate use of the component to me.

Reply:
I guess, it is more of... if the `RateByServiceSampler` is used to set the sampling priority (compared with setting `span.sampled`).

Which right now happens if `RateByServiceSampler` is set as the `post_sampler`/`priority_sampler` in here, *or* if the `RateByServiceSampler` is used by the `RuleSampler`

Reply:
I guess I'm less clear on the purpose of `_dd.agent_psr`: by definition, what is this tag supposed to represent? And how is its value used?

Reply:
Mostly for debugging, when you look at a raw trace in the backend you'll be able to see that the agent said to use sample rate of 0.2 to determine if this span should be kept or not.

Reply:
Should be addressed now.